From ce3633ccf3a2b74139c97aaa13743c044b98152f Mon Sep 17 00:00:00 2001
From: stheinan <stheinan@maxlinear.com>
Date: Mon, 1 Jul 2024 19:04:15 +0530
Subject: [PATCH] WLANRTSYS-62886 Limiting the Max MLDs to 15 and Max 6G MLD
 Vaps to 12

Issue:
Due to the limitation of max MPDU size which is 2304bytes
we cannot exceed this limit for Beacon , For every MLD vaps added
in 6G the additional RNR and MBSSID is added for the non transmitted
vaps in the single beacon, if more than 12 6G MLD vaps is created then
the Beacon size in 6G will exceed the limit of 2304.

Fix:
Limit the max MLD vaps in 6G to 12.

FIXUP to WLANRTSYS-60843 MLO changes: AP MLD configuration
---
 src/ap/hostapd.c |  6 ++++--
 src/ap/mxl_mld.c | 14 ++++++++++++++
 src/ap/mxl_mld.h |  3 ++-
 3 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 294fdb424..158b9df24 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3438,7 +3438,8 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 	hapd = iface->bss[0];
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
-	mxl_mld_setup_interfaces(iface->interfaces);
+	if(mxl_mld_setup_interfaces(iface->interfaces))
+		goto fail;
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
 	hostapd_tx_queue_params(hapd);
@@ -4387,7 +4388,8 @@ static int hostapd_reconf_add_bss(struct hostapd_iface *iface, struct hostapd_co
 	idx = iface->num_bss - 1;
 	hostapd_flush_old_stations(iface->bss[idx], WLAN_REASON_PREV_AUTH_NOT_VALID);
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
-	mxl_mld_setup_interfaces(iface->interfaces);
+	if(mxl_mld_setup_interfaces(iface->interfaces))
+		return -1;
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
 	if (start_ctrl_iface_bss(iface->bss[idx])){
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 2f043cdd3..51bd6604c 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -348,6 +348,7 @@ int mxl_mld_setup_interfaces(struct hapd_interfaces *interfaces)
 {
 	struct p_ap_mld *p_ap_mld = NULL;
 	struct hostapd_data *hapd1 = NULL, *hapd2 = NULL;
+	int num6ghz_mld_vaps=0;
 
 	if (interfaces->count <= 1)
 		return 0;
@@ -361,6 +362,19 @@ int mxl_mld_setup_interfaces(struct hapd_interfaces *interfaces)
 					!hapd1->beacon_set_done &&
 					hapd2 && hapd2->iface->set_freq_done &&
 					!hapd2->beacon_set_done) {
+				if((hapd1 && hapd1->iconf && is_6ghz_op_class(hapd1->iconf->op_class)) || (hapd2 && hapd2->iconf && is_6ghz_op_class(hapd2->iconf->op_class))) {
+					num6ghz_mld_vaps++;
+					if(num6ghz_mld_vaps > MAX_SUPPORTED_6G_MLDS) {
+						wpa_printf(MSG_ERROR, "Max MLD vap in 6G reached, MLD Creation failed");
+						if(hapd1 && hapd1->iface) {
+							mxl_hostapd_remove_bss(hapd1->iface, hapd1->conf->bss_idx);
+						}
+						if(hapd2 && hapd2->iface) {
+							mxl_hostapd_remove_bss(hapd2->iface, hapd2->conf->bss_idx);
+						}
+						return -1;
+					}
+				}
 				if (mxl_mld_link_setup(p_ap_mld)) {
 					wpa_printf(MSG_ERROR, "MLD setup failed for mld id %d", i);
 					return -1;
diff --git a/src/ap/mxl_mld.h b/src/ap/mxl_mld.h
index 334b2ae2f..94efbdda5 100644
--- a/src/ap/mxl_mld.h
+++ b/src/ap/mxl_mld.h
@@ -14,7 +14,8 @@
 #include "ap/ap_config.h"
 #include "ap/sta_info.h"
 
-#define MAX_SUPPORTED_MLDS 30
+#define MAX_SUPPORTED_MLDS 15
+#define MAX_SUPPORTED_6G_MLDS 12
 #define ML_BSS_LIMIT_PER_RADIO 16
 #define INVALID_MLD_ID 31
 /* IE ID + Len + Ext ID */
-- 
2.43.0

